File: rust/examples/ws_stream_mcap/src/playback_source.rs
use std::time::Duration;

use anyhow::Result;
use foxglove::{WebSocketServerHandle, websocket::PlaybackStatus};

/// A timestamp in nanoseconds since epoch.
pub type Nanoseconds = u64;

/// A data source that supports playback control with play/pause, seek, and variable speed.
///
/// Implementations are responsible for:
/// - Tracking playback state (playing/paused/ended) and current position
/// - Pacing message delivery according to timestamps and playback speed
/// - Logging messages to channels and broadcasting time updates to the server
pub trait PlaybackSource {
    /// Returns the (start, end) time bounds of the data.
    ///
    /// Determining this is dependent on the format of data you are loading.
    fn time_range(&self) -> (Nanoseconds, Nanoseconds);

    /// Sets the playback speed multiplier (e.g., 1.0 for real-time, 2.0 for double speed).
    ///
    /// Called by a ServerListener when it receives a PlaybackControlRequest from Foxglove
    fn set_playback_speed(&mut self, speed: f32);

    /// Begins or resumes playback.
    ///
    /// Called by a ServerListener when it receives a PlaybackControlRequest from Foxglove
    fn play(&mut self);

    /// Pauses playback.
    ///
    /// Called by a ServerListener when it receives a PlaybackControlRequest from Foxglove
    fn pause(&mut self);

    /// Seeks to the specified timestamp.
    ///
    /// Called by a ServerListener when it receives a PlaybackControlRequest from Foxglove
    fn seek(&mut self, log_time: Nanoseconds) -> Result<()>;

    /// Returns the current playback status.
    ///
    /// Used to send a PlaybackState to Foxglove
    fn status(&self) -> PlaybackStatus;

    /// Returns the current playback position.
    ///
    /// Used to send a PlaybackState to Foxglove
    fn current_time(&self) -> Nanoseconds;

    /// Returns the current playback speed multiplier.
    ///
    /// Used to send a PlaybackState to Foxglove
    fn playback_speed(&self) -> f32;

    /// Logs the next message for playback if it's ready, or returns a wall duration to wait
    /// (accounting for the current playback speed). This should be called by your main playback
    /// loop.
    ///
    /// Returns `Ok(Some(duration))` if the caller should sleep before calling again.
    /// Returns `Ok(None)` if a message was logged or playback is not active.
    ///
    /// The caller should sleep outside of any lock to allow control requests to be processed.
    /// This method also broadcasts time updates via `server.broadcast_time()`.
    fn log_next_message(&mut self, server: &WebSocketServerHandle) -> Result<Option<Duration>>;
}

File: python/foxglove-sdk-examples/ws-playback-control-mcap/playback_source.py
"""Abstract base class for a data source that supports playback control.

Implement this with your own data format, then reuse the ServerListener
and main-loop structure from main.py in your own player application.
"""

from abc import ABC, abstractmethod

from foxglove.websocket import PlaybackStatus, WebSocketServer


class PlaybackSource(ABC):
    """A data source that supports playback control with play/pause, seek, and variable speed.

    Implementations are responsible for:
    - Tracking playback state (playing/paused/ended) and current position
    - Pacing message delivery according to timestamps and playback speed
    - Logging messages to channels and broadcasting time updates to the server
    """

    @abstractmethod
    def time_range(self) -> tuple[int, int]:
        """Returns the (start, end) time bounds of the data in nanoseconds.

        Determining this is dependent on the format of data you are loading.
        """
        ...

    @abstractmethod
    def play(self) -> None:
        """Begins or resumes playback.

        Called by a ServerListener when it receives a PlaybackControlRequest from Foxglove.
        """
        ...

    @abstractmethod
    def pause(self) -> None:
        """Pauses playback.

        Called by a ServerListener when it receives a PlaybackControlRequest from Foxglove.
        """
        ...

    @abstractmethod
    def seek(self, log_time: int) -> None:
        """Seeks to the specified timestamp in nanoseconds.

        Called by a ServerListener when it receives a PlaybackControlRequest from Foxglove.
        """
        ...

    @abstractmethod
    def set_playback_speed(self, speed: float) -> None:
        """Sets the playback speed multiplier (e.g., 1.0 for real-time, 2.0 for double speed).

        Called by a ServerListener when it receives a PlaybackControlRequest from Foxglove.
        """
        ...

    @abstractmethod
    def status(self) -> PlaybackStatus:
        """Returns the current playback status.

        Used to send a PlaybackState to Foxglove.
        """
        ...

    @abstractmethod
    def current_time(self) -> int:
        """Returns the current playback position in nanoseconds.

        Used to send a PlaybackState to Foxglove.
        """
        ...

    @abstractmethod
    def playback_speed(self) -> float:
        """Returns the current playback speed multiplier.

        Used to send a PlaybackState to Foxglove.
        """
        ...

    @abstractmethod
    def log_next_message(self, server: WebSocketServer) -> float | None:
        """Logs the next message for playback if it's ready, or returns a duration to wait.

        Returns seconds to sleep if the caller should wait before calling again,
        or None if a message was logged or playback is not active.

        The caller should sleep outside of any lock to allow control requests to be processed.
        This method also broadcasts time updates via server.broadcast_time().
        """
        ...
